### PR TITLE
Device manager

### PR DESCRIPTION
**Pull request by Tim-0731-Hzt**

### CODE REVIEW

File: src/libs/kata-sys-util/src/lib.rs [hunk 1]
```diff
@@ -31,3 +33,34 @@ macro_rules! eother {
         std::io::Error::new(std::io::ErrorKind::Other, format!($fmt, $($arg)*))
     })
 }
+
+
+pub struct HexSlice<'a>(&'a [u8]);
```
Comment (quanweiZhou) [inline src/libs/kata-sys-util/src/lib.rs:38]: It seems useless to define `struct HexSlice`, the `rand::RandomBytes' is already supported.

File: src/runtime-rs/crates/resource/src/device/device/config.rs [hunk 2]
```diff
@@ -0,0 +1,48 @@
+use hypervisor::IoLimits;
+use std::collections::HashMap;
+
+/// DeviceInfo is an embedded type that contains device data common to all types of devices.
+#[derive(Default, Debug, Clone)]
+pub struct DeviceInfo {
```
Comment (quanweiZhou) [inline src/runtime-rs/crates/resource/src/device/device/config.rs:6]: Suggest adding functions for new devices, such as `new_block_devid()`.​

File: src/runtime-rs/crates/resource/src/device/manager.rs [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+                return Some(Arc::clone(dev));
+            }
+        }
+        None
+    }
+
+    fn new_device_id(&self) -> Result<String> {
+        for _ in 0..5 {
+            let rand_bytes = rand::RandomBytes::new(8).bytes;
+            let id = format!("{:x}", HexSlice::new(&rand_bytes));
```
Comment (quanweiZhou) [inline src/runtime-rs/crates/resource/src/device/manager.rs:197]: It seems useless to define `HexSlice`, the `RandomBytes` already support `to hex`.

File: src/runtime-rs/crates/resource/Cargo.toml [hunk 4]
```diff
@@ -17,6 +17,7 @@ netlink-packet-route = "0.13.0"
 nix = "0.24.1"
 rand = "^0.7.2"
 rtnetlink = "0.11.0"
+rust-ini = "0.18.0"
```
Comment (quanweiZhou) [inline src/runtime-rs/crates/resource/Cargo.toml:20]: ​Why need this dependency?​

File: src/runtime-rs/crates/resource/src/device/device/mod.rs [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+#[async_trait]
+pub trait Device: Send + Sync {
+    async fn attach(&mut self, h: &dyn Hypervisor) -> Result<()>;
+    async fn detach(&mut self, h: &dyn Hypervisor) -> Result<()>;
+
+    async fn device_id(&self) -> &str;
+    async fn get_device_info(&self) -> Result<DeviceInfo>;
+    async fn get_major_minor(&self) -> (i64, i64);
+    async fn get_host_path(&self) -> &str;
+    async fn get_bdf(&self) -> Option<&String>;
```
Comment (quanweiZhou) [inline src/runtime-rs/crates/resource/src/device/device/mod.rs:22]: I think the trait function `get_bdf` is not a common device operation.

File: src/runtime-rs/crates/resource/src/device/driver/vfio.rs [hunk 6]
```diff
@@ -0,0 +1 @@
+
```
Comment (quanweiZhou) [inline src/runtime-rs/crates/resource/src/device/driver/vfio.rs:1]: Unused file.

File: src/runtime-rs/crates/resource/src/device/manager.rs [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+/// VirtioMmio indicates block driver is virtio-mmio based
+pub const VIRTIO_MMIO: &str = "virtio-mmio";
+pub const VIRTIO_BLOCK: &str = "virtio-blk";
+pub const VFIO: &str = "vfio";
+pub const KATA_MMIO_BLK_DEV_TYPE: &str = "mmioblk";
+pub const KATA_BLK_DEV_TYPE: &str = "blk";
+type ArcBoxDevice = Arc<Mutex<Box<dyn Device>>>;
+
+pub struct DeviceManager {
+    block_driver: String,
```
Comment (quanweiZhou) [inline src/runtime-rs/crates/resource/src/device/manager.rs:21]: I think it's better to add a `DeviceConfig` and include `block_driver`.

File: src/runtime-rs/crates/resource/src/device/manager.rs [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+        } else {
+            return Err(anyhow!(
+                "device with specified ID hasn't been created. {}",
+                id
+            ));
+        }
+        Ok(())
+    }
+
+    async fn _detach_device(&self, id: &str, h: &dyn Hypervisor) -> Result<()> {
```
Comment (quanweiZhou) [inline src/runtime-rs/crates/resource/src/device/manager.rs:103]: Lost detached process?

File: src/runtime-rs/crates/resource/src/device/manager.rs [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+    ) -> Result<AgentDevice> {
+        // Safe because we just attached the device
+        let dev = self.get_device_by_id(&device_info.id);
+        let base_info = dev.unwrap().lock().await.get_device_info().await?;
+        let mut device = AgentDevice {
+            container_path: device_info.container_path.clone(),
+            ..Default::default()
+        };
+
+        match self.get_block_driver().await {
```
Comment (quanweiZhou) [inline src/runtime-rs/crates/resource/src/device/manager.rs:63]: I think agent device ~= block device

Comment (Tim-0731-Hzt) [reply src/runtime-rs/crates/resource/src/device/manager.rs:63]: agent device is needed by create container request

File: src/runtime-rs/crates/resource/src/manager_inner.rs [hunk 10]
```diff
@@ -33,6 +37,7 @@ pub(crate) struct ResourceManagerInner {
     hypervisor: Arc<dyn Hypervisor>,
     network: Option<Arc<dyn Network>>,
     share_fs: Option<Arc<dyn ShareFs>>,
+    device: Arc<Mutex<DeviceManager>>,
```
Comment (YushuoEdge) [inline src/runtime-rs/crates/resource/src/manager_inner.rs:40]: I wonder whether other types of containers will use the `DeviceManager`. If not, I think maybe we can create `VmResource` here, just like `CgroupsResource `, and move the `DeviceManager` to the` VmResource`. By this way we can also manage some other resources like cpu/mem in `VmResource`.

**Pull request closed without merging**